sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,207 @@
+            }
+        }
+
+        private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
+        {
+            using DiagnosticScope scope = ClientDiagnostics.CreateScope($"{OperationTypeName}.UpdateStatus");
+
+            foreach (KeyValuePair<string, string> attribute in ScopeAttributes)
+            {
+                scope.AddAttribute(attribute.Key, attribute.Value);
+            }
+
+            scope.Start();
+
+            try
+            {
+                return await UpdateStateAsync(async, cancellationToken).ConfigureAwait(false);
+            }
+            catch (RequestFailedException e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+            catch (Exception e)
+            {
+                var requestFailedException = new RequestFailedException("Unexpected failure.", e);
+                scope.Failed(requestFailedException);
+                throw requestFailedException;
+            }
+        }
+
+        private async ValueTask<Response> UpdateStateAsync(bool async, CancellationToken cancellationToken)
+        {
+            var state = await _operation.UpdateStateAsync(async, cancellationToken).ConfigureAwait(false);
+
+            lock (SetStateLock)
+            {
+                RawResponse = state.RawResponse;
+
+                if (state.Succeeded == true)
+                {
+                    Value = state.Value;
+                    HasCompleted = true;
+                }
+                else if (state.Succeeded == false)
+                {
+                    OperationFailedException = state.OperationFailedException;
+                    HasCompleted = true;
+
+                    throw OperationFailedException;
+                }
+            }
+
+            return state.RawResponse;
+        }
+    }
+
+    internal interface IOperation<TResult>
+    {
+        ValueTask<OperationState<TResult>> UpdateStateAsync(bool async, CancellationToken cancellationToken);
+    }
+
+    internal readonly struct OperationState<TResult>
+    {
+        private OperationState(Response rawResponse, bool? succeeded, TResult value, RequestFailedException operationFailedException)
+        {
+            RawResponse = rawResponse;
+            Succeeded = succeeded;
+            Value = value;
+            OperationFailedException = operationFailedException;
+        }
+
+        public Response RawResponse { get; }
+
+        public bool? Succeeded { get; }
+
+        public TResult Value { get; }
+
+        public RequestFailedException OperationFailedException { get; }
+
+        public static OperationState<TResult> Success(Response rawResponse, TResult value) =>
+            new OperationState<TResult>(rawResponse, true, value, default);
+
+        public static OperationState<TResult> Failure(Response rawResponse, RequestFailedException operationFailedException) =>
```

Comment:
Consider making the operationFailedException parameter defaulted. I think we can get a pretty reasonable default experience.